### PR TITLE
Category

### PR DESCRIPTION
**Pull request by Olena-Pro**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/model/Category.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "name",
+            nullable = false)
+    private String name;
+    @Column(name = "description")
+    private String description;
+    @Column(name = "is_deleted",
+            nullable = false)
```
Comment (koval-kn) [inline src/main/java/mate/academy/bookstore/model/Category.java]: not critical but in this case explicitly specifying column name is redundant as Hibernate will automatically map it (for example field name `isDeleted` will be mapped to `is_deleted` implicitly as default)
```suggestion
    @Column(nullable = false)
    private String name;
    private String description;
    @Column(nullable = false)
```

File: src/main/java/mate/academy/bookstore/security/JwtUtil.java [hunk 2]
```diff
@@ -37,7 +36,7 @@ public boolean isValidToken(String token) {
                     .setSigningKey(secret)
                     .build()
                     .parseClaimsJws(token);
-            return !getClaimsFromToken(token, Claims::getExpiration).before(new Date());
+            return ! getClaimsFromToken(token, Claims::getExpiration).before(new Date());
```
Comment (koval-kn) [inline src/main/java/mate/academy/bookstore/security/JwtUtil.java]: extract `new Date()` into `today` variable

File: src/main/java/mate/academy/bookstore/service/BookService.java [hunk 3]
```diff
@@ -15,8 +16,10 @@ public interface BookService {
 
     void deleteById(Long id);
 
-    BookDto update(Long id,CreateBookRequestDto requestDto);
+    BookDto update(Long id, CreateBookRequestDto requestDto);
 
     List<BookDto> search(BookSearchParameters params);
 
+    List<BookWithoutCategoryDto> findAllByCategoryId(Long id);
+
```
Comment (koval-kn) [inline src/main/java/mate/academy/bookstore/service/BookService.java]:
```suggestion
```

File: src/main/resources/db/changelog/changes/10-create-category-table.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
```
Comment (koval-kn) [inline src/main/resources/db/changelog/changes/10-create-category-table.yaml]: description can be long
```suggestion
                  type: text
```

File: target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/BookMapperImpl.java
Comment (koval-kn) [inline target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/BookMapperImpl.java:1]: target folder should not be included to git

File: target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/CategoryMapperImpl.java
Comment (koval-kn) [inline target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/CategoryMapperImpl.java:1]: target folder should not be included to git

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/CategoryMapperImpl.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import mate.academy.bookstore.model.Category;
+import org.springframework.stereotype.Component;
+
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor",
+    date = "2024-05-06T13:16:00+0300",
+    comments = "version: 1.5.5.Final, compiler: javac, environment: Java 19.0.2 (Oracle Corporation)"
+)
+@Component
+public class CategoryMapperImpl implements CategoryMapper {
```
Comment (Elena-Bruyako) [inline target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/CategoryMapperImpl.java:16]: These classes shouldn't be visible, add target to gitignore

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category menegment", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
@RequestMapping("/categories")
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    @Operation(summary = "Delete a category by id", description = "Delete a category by id")
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Update category", description = "Update a category by id")
+    public CategoryDto updateCategory(@PathVariable Long id,
+                                      @RequestBody CategoryRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
                                      @RequestBody @Valid CategoryRequestDto requestDto) {
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return categoryService.update(id, requestDto);
+    }
+
+    @GetMapping("/{id}/books")
+    @Operation(summary = "Get books by category",
+            description = "Get a list of books by category id")
+    public List<BookWithoutCategoryDto> getBooksByCategoryId(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
    public List<BookWithoutCategoryDto> getBooksByCategoryId(@PathVariable @Positive Long id) {
```
Check other id 
Add `@Validated` under the controller

File: src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java [hunk 9]
```diff
@@ -20,4 +21,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java]: For String is better use `@NotBlank`

File: src/main/java/mate/academy/bookstore/dto/category/CategoryRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.bookstore.dto.category;
+
+public record CategoryRequestDto(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/category/CategoryRequestDto.java:5]: Add validation

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 11]
```diff
@@ -40,4 +45,10 @@ public class Book {
 
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java]: Bad approach to use EAGER

File: src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        Category category = categoryMapper.toModel(requestDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryRequestDto requestDto) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find category by id: " + id)
+        );
+        Category categoryWithId = categoryMapper.toModel(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java]: You can create in CategoryMapper method: 
`void updateCategoryFromDto(CategoryRequestDto requestDto, @MappingTarget Category category);`
and use it here

File: src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml]: fix the mistake in the file name

File: src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  - changeSet:
+      id: create-books_categories-table
+      author: Olena
+      changes:
+        - createTable:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
+                  type: bigint
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml]: constraints ?

File: src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+              - column:
+                  name: category_id
+                  type: bigint
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-books-categouries-table.yaml]: constraints ?

File: src/main/resources/db/changelog/changes/12-add-fks-in-books_categories-table.yaml [hunk 16]
```diff
@@ -0,0 +1,21 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-forign-keys-in-books-categories
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/12-add-fks-in-books_categories-table.yaml]: Don't need to create addForeignKeyConstraint here, you can add them in the create-books_categories-table changelog

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category menegment", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java:33]:
```suggestion
@RequestMapping("/categories")
@Validated
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @GetMapping
+    @Operation(summary = "Find all categories", description = "Get a list of all categories")
+    public List<CategoryDto> getAll() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
    public List<CategoryDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 19]
```diff
@@ -40,4 +45,10 @@ public class Book {
 
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java]: ManyToMany has default FetchType.LAZY

File: src/main/resources/db/changelog/changes/11-create-books-categories-table.yaml [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+              - id
+            constraintName: fk_book_id
+        - addForeignKeyConstraint:
+            baseTableName: books_categories
+            baseColumnNames:
+              - category_id
+            referencedTableName: categories
+            referencedColumnNames:
+              - id
+            constraintName: fk_category_id
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-books-categories-table.yaml:34]:
```suggestion
            constraintName: fk_category_id
            
```

File: src/main/resources/db/changelog/changes/10-create-category-table.yaml [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+                    nullable: false
+              - column:
+                  name: description
+                  type: text
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-category-table.yaml:29]:
```suggestion
                    nullable: false
                    
```

File: target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/BookMapperImpl.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+import mate.academy.bookstore.model.Book;
+import org.springframework.stereotype.Component;
+
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor",
+    date = "2024-05-06T13:16:00+0300",
+    comments = "version: 1.5.5.Final, compiler: javac, environment: Java 19.0.2 (Oracle Corporation)"
+)
+@Component
+public class BookMapperImpl implements BookMapper {
```
Comment (Elena-Bruyako) [inline target/generated-sources/annotations/mate/academy/bookstore/mapper/impl/BookMapperImpl.java:18]: Please, add these to gitignore 
target/
!**/src/main/**/target/
!**/src/test/**/target/

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

**Review by sarakhmen — APPROVED** [review]
good job!

**Pull request merged**